### PR TITLE
Do not publish zero press counter value after device's reset

### PR DESCRIPTION
**Pull request by KraPete**
## Summary by CodeRabbit

- **New Features**
	- Added support for a new register type, `press_counter`.
	- Introduced callbacks for device connection state changes in multiple classes.
	- Enhanced device configuration schema with clearer property definitions and new enumerated values.

- **Bug Fixes**
	- Updated `wb-mqtt-serial` package to prevent erroneous zero press counter values after device resets.

- **Refactor**
	- Simplified method signatures and removed unnecessary parameters related to device connection state management.
	- Streamlined connection state handling across various components.
	- Updated JSON templates to reflect changes in input handling, specifically changing `reg_type` from `input` to `press_counter` for multiple devices.

- **Tests**
	- Enhanced testing framework to allow precise control over timing in serial communication tests.

- **Documentation**
	- Improved clarity and comprehensiveness of the README documentation, including updates on supported protocols and device configuration options.

### CODE REVIEW

File: src/serial_client_events_reader.h [hunk 1]
```diff
@@ -67,4 +64,5 @@ class TSerialClientEventsReader
     void OnEnabledEvent(uint8_t slaveId, uint8_t type, uint16_t addr, bool res);
     void ClearReadErrors(TRegisterCallback callback);
     void ReadEventsFailed(const std::string& errorMessage, TRegisterCallback registerCallback);
+    void OnDeviceConnectionStateChanged(PSerialDevice device);
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_client_events_reader.h:67]: useless

File: src/serial_client_register_poller.h [hunk 2]
```diff
@@ -36,27 +36,23 @@ class TSerialClientRegisterPoller
...
-                              TRegisterCallback callback,
-                              TDeviceCallback deviceConnectionStateChangedCallback);
-    void DeviceDisconnected(PSerialDevice device, std::chrono::steady_clock::time_point currentTime);
+                              TRegisterCallback callback);
 
 private:
     void ScheduleNextPoll(PPollableDevice device);
     std::chrono::steady_clock::time_point GetDeadline(bool lowPriorityRateLimitIsExceeded,
                                                       const util::TSpentTimeMeter& spentTime) const;
+    void OnDeviceConnectionStateChanged(PSerialDevice device);
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_client_register_poller.h:55]: useless

File: src/devices/modbus_device.cpp [hunk 3]
```diff
@@ -10,7 +10,8 @@ namespace
                                               {Modbus::REG_HOLDING_MULTI, "holding_multi", "value", U16},
                                               {Modbus::REG_COIL, "coil", "switch", U8},
                                               {Modbus::REG_DISCRETE, "discrete", "switch", U8, true},
-                                              {Modbus::REG_INPUT, "input", "value", U16, true}});
+                                              {Modbus::REG_INPUT, "input", "value", U16, true},
+                                              {Modbus::REG_INPUT, "press_counter", "value", U16, true}});
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/devices/modbus_device.cpp:14]: useless

File: src/serial_client_register_poller.cpp [hunk 4]
```diff
@@ -254,26 +229,18 @@ TPollResult TSerialClientRegisterPoller::OpenPortCycle(TPort& port,
...
+void TSerialClientRegisterPoller::OnDeviceConnectionStateChanged(PSerialDevice device)
 {
-    auto range = Devices.equal_range(device);
-    for (auto it = range.first; it != range.second; ++it) {
-        it->second->RescheduleAllRegisters(currentTime);
+    if (device->GetConnectionState() == TDeviceConnectionState::DISCONNECTED) {
+        auto range = Devices.equal_range(device);
+        for (auto it = range.first; it != range.second; ++it) {
+            it->second->RescheduleAllRegisters();
+        }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_client_register_poller.cpp:243]: useless

File: src/serial_device.cpp [hunk 5]
```diff
@@ -273,6 +269,27 @@ void TSerialDevice::SetLastReadTime(std::chrono::steady_clock::time_point readTi
...
+    ConnectionState = state;
+    if (state == TDeviceConnectionState::CONNECTED) {
+        LOG(Info) << "device " << ToString() << " is connected";
+    } else {
+        LOG(Warn) << "device " << ToString() << " is disconnected";
+    }
+    for (auto& callback: ConnectionStateChangedCallbacks) {
+        callback(shared_from_this());
+    }
+}
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_device.cpp:291]: Overengineering

File: src/serial_device.h [hunk 6]
```diff
@@ -231,6 +233,8 @@ class TSerialDevice: public std::enable_shared_from_this<TSerialDevice>
     std::chrono::steady_clock::time_point GetLastReadTime() const;
     void SetLastReadTime(std::chrono::steady_clock::time_point readTime);
 
+    void AddOnConnectionStateChangedCallback(TDeviceCallback callback);
+
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_device.h:237]: useless

File: src/serial_client_events_reader.cpp [hunk 7]
```diff
@@ -318,24 +314,33 @@ void TSerialClientEventsReader::OnEnabledEvent(uint8_t slaveId, uint8_t type, ui
...
+                        static_cast<uint8_t>(dev->SlaveId),
+                        static_cast<uint16_t>(GetUint32RegisterAddress(reg->GetAddress())),
+                        ToEventRegisterType(static_cast<Modbus::RegisterType>(reg->Type))};
+                    Regs[regDesc].push_back(reg);
+                }
+            }
         }
+        dev->AddOnConnectionStateChangedCallback(
+            [this](PSerialDevice device) { OnDeviceConnectionStateChanged(device); });
     }
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_client_events_reader.cpp:334]: useless

File: src/serial_device.cpp [hunk 8]
```diff
@@ -273,6 +269,27 @@ void TSerialDevice::SetLastReadTime(std::chrono::steady_clock::time_point readTi
     LastReadTime = readTime;
 }
 
+void TSerialDevice::AddOnConnectionStateChangedCallback(TSerialDevice::TDeviceCallback callback)
+{
+    ConnectionStateChangedCallbacks.push_back(callback);
+}
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_device.cpp:275]: useless

File: src/serial_port_driver.cpp [hunk 9]
```diff
@@ -395,3 +434,23 @@ bool TDeviceChannel::HasValuesOfAllRegisters() const
...
+            try {
+                if (r->GetValue().Get<uint16_t>() == 0) {
+                    return true;
+                }
+            } catch (const TRegisterValueException&) {
+            }
+        }
+    }
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_port_driver.cpp:451]: useless

File: src/serial_client.h [hunk 10]
```diff
@@ -24,19 +24,15 @@ class TSerialClientRegisterAndEventsReader: public util::TNonCopyable
...
 
     TSerialClientRegisterAndEventsReader(const std::list<PSerialDevice>& devices,
                                          std::chrono::milliseconds readEventsPeriod,
                                          util::TGetNowFn nowFn,
                                          size_t lowPriorityRateLimit = std::numeric_limits<size_t>::max());
 
-    void ClosedPortCycle(std::chrono::steady_clock::time_point currentTime,
-                         TRegisterCallback regCallback,
-                         TDeviceCallback deviceConnectionStateChangedCallback);
+    void ClosedPortCycle(std::chrono::steady_clock::time_point currentTime, TRegisterCallback regCallback);
```
[the remark this answers is not in the document]
Comment (KraPete) [reply src/serial_client.h:33]: useless

File: debian/changelog [hunk 11]
```diff
@@ -1,3 +1,9 @@
+wb-mqtt-serial (2.146.0) stable; urgency=medium
+
+  * Do not publish zero press counter value after device's reset
+
+ -- Petr Krasnoshchekov <petr.krasnoshchekov@wirenboard.com>  Thu, 31 Oct 2024 12:25:29 +0500
```
[the remark this answers is not in the document]
Comment (KraPete) [reply debian/changelog]: fixed

**Comment by KraPete** [comment]
@pgasheev Посмотри, пожалуйста, правильно ли я пометил каналы, которые сбрасываются в 0 после перезагрузки? Возможно, что-то забыл или слишком много пометил

**Review by sikmir — APPROVED** [review]

**Pull request merged**